src/test/scala/chiselTests/TypeAliasSpec.scala

```diff
@@ -0,0 +1,321 @@
+      "Attempted to redeclare an existing bundle type alias 'DifferentBundle' with a new bundle structure"
+    )
+    msg should include("The alias was previously defined as:")
+    msg should include("@[src/test/scala/chiselTests/TypeAliasSpec.scala")
+  }
+
+  "Bundles with unsanitary names" should "properly sanitize" in {
+    class Test extends Module {
+      class FooBundle extends Bundle with HasTypeAlias {
+        // Sanitizes to '_'; the sanitized alias needs to be used in the aliasing algorithm
+        // instead of the direct user-defined alias.
+        override def aliasName = Some(BundleAlias(""))
+
+        val x = UInt(8.W)
+        val y = UInt(8.W)
+      }
+
+      val io = IO(new Bundle {
+        val in = Input(new FooBundle)
+        val out = Output(new FooBundle)
+      })
+
+      val w = Wire(new FooBundle)
+
+      w :#= io.in
+      io.out :#= w
+    }
+
+    val chirrtl = ChiselStage.emitCHIRRTL(new Test)
+
+    chirrtl should include("type _ = { x : UInt<8>, y : UInt<8>}")
+    chirrtl should include("output io : { flip in : _, out : _}")
+    chirrtl should include("wire w : _")
+  }
+
+  "Bundle type aliases overriding an existing FIRRTL type" should "error" in {
+    // Special keywords/types specified in the FIRRTL spec.
+    // These result in parser errors and should not be allowed by Chisel
+    // Duplicated from the same list in the `internal` package object, which is private to chisel3
+    val firrtlKeywords = Seq(
+      "FIRRTL",
+      "Clock",
+      "UInt",
+      "Reset",
+      "AsyncReset",
+      "Analog",
+      "Probe",
+      "RWProbe",
+      "version",
+      "type",
+      "circuit",
+      "parameter",
+      "input",
+      "output",
+      "extmodule",
+      "module",
+      "intmodule",
+      "intrinsic",
+      "defname",
+      "const",
+      "flip",
+      "reg",
+      "smem",
+      "cmem",
+      "mport",
+      "define",
+      "attach",
+      "inst",
+      "of",
+      "reset",
+      "printf",
+      "skip",
+      "node"
+    )
+
+    // Prevent statements like type Clock = { ... }
+    firrtlKeywords.map { tpe =>
+      (the[ChiselException] thrownBy extractCause[ChiselException] {
+        class Test(val firrtlType: String) extends Module {
+          class FooBundle extends Bundle with HasTypeAlias {
+            override def aliasName = Some(BundleAlias(firrtlType))
+
+            val x = UInt(8.W)
+          }
+
+          val io = IO(new Bundle {
+            val in = Input(new FooBundle)
+            val out = Output(new FooBundle)
+          })
+
+          io.out.x :#= io.in.x
+        }
+
+        val args = Array("--throw-on-first-error", "--full-stacktrace")
+        val chirrtl = ChiselStage.emitCHIRRTL(new Test(tpe), args)
+      }).getMessage should include(
+        s"Attempted to override a FIRRTL keyword '$tpe' with a bundle type alias. Chisel does not automatically disambiguate aliases using these keywords at this time."
+      )
+    }
+  }
```

Comment:
Can you add a test where you mix `HasTypeAlias` into a `Record`? From my reading of the code it's legal but not correctly implemented [yet].

Reply:
please show a test for this, can reimplement one or more of the above tests